README.md
# dos-c

## Introduction
This diskette contains DOS-C version 0.91b for PC/XT/AT/AT386/AT486
computers.  This is the August 28, 1996 release in an ongoing project.  
There is a companion book, "The FreeDOS Kernel" (ISBN: 0-87930-436-7) 
published by R&D Books, an imprint of Miller Freeman of Lawrence, 
Kansas (USA) and distributed in the USA and Canada by Publishers Group West. 
See below for addresses.

DOS-C is available by anonymous ftp at ftp.iop.com in /pub/freedos and via 
World Wide Web at http://www.iop.com/~patv.

DOS-C is also available through the FreeDOS project.  An ftp site is 
available for downloading Free-DOS at sunsite.unc.edu in 
/pub/micro/pc-stuff/freedos.

## Agreement
All users of DOS-C must accept the disclaimer of warranty and license terms 
contained in the file "COPYING" in order to use it.  You may not, under 
any circumstance, use this operating system or any component without first 
reading and agreeing to the terms of the license.

## Bug Reports
If you have found a bug, think you have found a bug, or would just like to 
make a suggestion, send email to freedos@sunsite.unc.edu or  to the author 
below.  All suggestions or reports will receive the author's attention.  
Just have patience, as this is all uncompensated work and the authors need 
to make a living.

For this version, submit at https://github.com/pingleware/dos-c/issues

## Wiki

For tips and techniques relating to the build, installation, configuration, etc. at

    https://github.com/pingleware/dos-c/wiki
    

## Author
Pat Villani (patv@iop.com)

## Maintainer
PressPage Entertainment Inc. (https://presspage.info) [presspage.entertainment@gmail.com]

## Copyright
(c) Copyright 1995, 1996 by Pasquale J. Villani
All Rights Reserved.


## Addresses
R&D Books
1601 West 23rd Street, Suite 200
Lawrence, Kansas 66046
USA

Publishers Group West
P.O. Box 8843
Emeryvill, California 94662
USA

PressPage Enterainment Inc
P.O. Box
Gainesville, Florida


history.txt
DOS-C started in 1988 as an experiment in writing device drivers in C for 
Microsoft's MS-DOS.  Both block and character device drivers were written, 
along with special C data structures to match the MS-DOS request packet.  It 
was then recognized that using the same techniques, an operating system 
could be written that would take advantage of the C language features and 
would require much less time to develop than the traditional assembly 
language techniques.  Although UNIX had proven this earlier, it was not 
tried with a traditional pc operating system.

At this time, a minimal operating system using the device drivers written 
earlier along with a new 8086 interrupt API was developed.  It was called 
XDOS and proved to be a functional operating system.  This new operating 
system was used to develop booting techniques and a C library SDK was 
developed for it.

XDOS enhancements were started in 1989 and MS-DOS was chosen as the new API.  
A more advanced architecture was also developed.  This included the use of 
an IPL (intermediate program loader) to set up the operating environment 
prior to loading the operating system itself and reentrant system calls 
facilitating real-time applications.  This version, know as NSS-DOS, was 
completed and demonstrated in 1991.  As a result of these demonstrations, 
NSS was approached to supply source license for this operating system by a 
major defense contractor.  The only new requirement - it had to run on 68K 
processors.

This presented a new challenge.  Due to the MS-DOS model used for the API, 
NSS-DOS relied heavily on a segmented architecture.  To meet this challenge, 
a major redesign of NSS-DOS was undertaken.  New proprietary techniques were 
developed that allowed the same source to be compiled on a variety of hosts 
and with a wide range of compilers.  This new version, DOS/NT, was the result o
of this new project.  The kernel was redesigned as a micro kernel along with 
logical separation of the filesystem, memory and task managers.  A new DOS 
API was designed along with a new DOS SDK to guarantee portability.  
Additionally, all processor unique code was separated from the core 
functions.  The result is  the highly portable operating system that DOS/NT 
represents.

After a number of successful commercial applications, DOS/NT became part of
both dosemu and FreeDOS.

This version, DOS-C, is the subject of an upcoming book and is intended for 
binary redistribution, free of any royalty.  See the accompanying 
license.txt file for details.  If you have not received this file, contact 
me at the address below and I will send a copy to you.

The DOS-C kernel is also the FreeDOS kernel.  FreeDOS is a project designed 
to provide an alternative to MS-DOS, PC-DOS and DR-DOS that is freely 
available.  See the FreeDOS manifesto for details.

MS-DOS is a trademark of Microsoft Corporation.
UNIX is a trademark of USL, Inc.


Author
------
Pat Villani (patv@iop.com)

Copyright
---------
(c) Copyright 1995 by Pasquale J. Villani
All Rights Reserved.

License
-------
See COPYING in DOS-C root directory for license.


$Header:   C:/dos-c/doc/history.txv   1.3   29 Aug 1996 13:06:42   patv  $

$Log:   C:/dos-c/doc/history.txv  $
   
      Rev 1.3   29 Aug 1996 13:06:42   patv
   Bug fixes for v0.91b
   
      Rev 1.2   19 Feb 1996  3:08:18   patv
   Updated for v0.91
   
      Rev 1.1   01 Sep 1995 17:30:42   patv
   First GPL release
   
      Rev 1.0   05 Jul 1995 18:50:22   patv
   Initial revision.



hdr/pcb.h

/****************************************************************/
/*								*/
/*			       pcb.h				*/
/*								*/
/*	     Process Control and Interrupt data structures	*/
/*								*/
/*			 November 26, 1991			*/
/*								*/
/*			Copyright (c) 1995			*/
/*			Pasquale J. Villani			*/
/*			All Rights Reserved			*/
/*								*/
/* This file is part of DOS-C.					*/
/*								*/
/* DOS-C is free software; you can redistribute it and/or	*/
/* modify it under the terms of the GNU General Public License	*/
/* as published by the Free Software Foundation; either version	*/
/* 2, or (at your option) any later version.			*/
/*								*/
/* DOS-C is distributed in the hope that it will be useful, but	*/
/* WITHOUT ANY WARRANTY; without even the implied warranty of	*/
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See	*/
/* the GNU General Public License for more details.		*/
/*								*/
/* You should have received a copy of the GNU General Public	*/
/* License along with DOS-C; see the file COPYING.  If not,	*/
/* write to the Free Software Foundation, 675 Mass Ave,		*/
/* Cambridge, MA 02139, USA.					*/
/****************************************************************/


/* $Logfile*/
#ifdef MAIN
#ifdef VERSION_STRINGS
static BYTE *pcb_hRcsId = "$Header:   D:/dos-c/hdr/pcb.h_v   1.4   29 May 1996 21:25:16   patv  $";
# endif
#endif

/*
 * $Log:   D:/dos-c/hdr/pcb.h_v  $
 *	
 *	   Rev 1.4   29 May 1996 21:25:16   patv
 *	bug fixes for v0.91a
 *	
 *	   Rev 1.3   19 Feb 1996  3:15:32   patv
 *	Added NLS, int2f and config.sys processing
 *	
 *	   Rev 1.2   01 Sep 1995 17:35:44   patv
 *	First GPL release.
 *	
 *	   Rev 1.1   30 Jul 1995 20:43:50   patv
 *	Eliminated version strings in ipl
 *	
 *	   Rev 1.0   02 Jul 1995 10:39:48   patv
 *	Initial revision.
 */

/*									*/
/* interrupt handler structure definition				*/
/*									*/
typedef union
{
	UWORD	x;		/* access mode for ax, bx, etc.		*/
	struct
	{
		UBYTE	l;	/* access mode for al, bl, etc.		*/
		UBYTE	h;	/* access mode for ah, bh, etc.		*/
	} b;
} xreg;

/* maps compiler unique stacking order					*/
typedef struct
{
#ifdef __BORLANDC__
	UWORD bp, di, si, ds, es;
	xreg d, c, b, a;
#else /* MSC */
	UWORD es, ds;
	UWORD di, si, bp, sp;
	xreg b, d, c, a;
#endif
	UWORD ip, cs, flags;
} iregs;

/* Process control block for task switching				*/
typedef struct
{
	UWORD pc_ss;
	UWORD pc_sp;
	iregs	pc_regs;
} pcb;

/* For MSC, the following offsets must match the assembly process	*/
/* support offsets							*/
/* NOTE: Alignemnts must be set to 1 (-Zp1)				*/
/*	ss:	0							*/
/*	sp:	2							*/
/*	es:	4							*/
/*	ds:	6							*/
/*	di:	8							*/
/*	si:	10							*/
/*	bp:	12							*/
/*	sp:	14	NOTE: not used in this structure		*/
/*	bx:	16							*/
/*	dx:	18							*/
/*	cx:	20							*/
/*	ax:	22							*/
/*	ip:	24							*/
/*	cs:	26							*/
/*	flags:	28							*/
/*									*/
/* For Borland C, the following offsets must match the assembly process	*/
/* support offsets							*/
/*	ss:	0							*/
/*	sp:	2							*/
/*	bp:	4							*/
/*	di:	6							*/
/*	si:	8							*/
/*	ds:	10							*/
/*	es:	12							*/
/*	dx:	14							*/
/*	cx:	16							*/
/*	bx:	18							*/
/*	ax:	20							*/
/*	ip:	22							*/
/*	cs:	24							*/
/*	flags:	26							*/


#define	BP	bp
#define	DI	di
#define	SI	si
#define	DS	ds
#define	ES	es
#define	DX	d.x 
#define	CX	c.x 
#define	BX	b.x 
#define	AX	a.x
#define	DH	d.b.h 
#define	CH	c.b.h 
#define	BH	b.b.h 
#define	AH	a.b.h
#define	DL	d.b.l 
#define	CL	c.b.l 
#define	BL	b.b.l 
#define	AL	a.b.l
#define	IP	ip
#define	CS	cs
#define	FLAGS	flags

#define	FLG_ZERO	0x0040
#define FLG_CARRY	0x0001


